### PR TITLE
feat(@nguniversal/common): add experimental SSR project Clover

### PR DESCRIPTION
**Pull request by alan-agius4**
# @nguniversal/common/clover

- What if Angular SSR didn't require the complex `@nguniversal` and `@angular/platform-server` packages on boarding?
- What if we `Window is undefined` error was a thing of the past?
- What if you don't need multiple builds for an SSR/prerender application?
- What if an application shell can be generated without an extra build?

## Supported features

- [x] Inline critical CSS
- [x] State transfer
- [x] Re-use component styles generated on the server
- [x] i18n
- [x] Hybrid rendering
- [x] Server side rendering

# Planned features
- [ ] App-shell
- [ ] Pre-rendering
- [ ] Schematics

### Install the dependencies
```
npm install @nguniversal/common express --save
```

#### app/app.module.ts

```diff
  import { NgModule } from '@angular/core';
  import { BrowserModule } from '@angular/platform-browser';

  import { AppRoutingModule } from './app-routing.module';
  import { AppComponent } from './app.component';
  import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+ import { RendererModule, TransferHttpCacheModule } from '@nguniversal/common/clover';

  @NgModule({
    declarations: [
      AppComponent
    ],
    imports: [
-     BrowserModule,
+     BrowserModule.withServerTransition({
+      appId: 'myapp',
+     }),
+     RendererModule.forRoot(),
+     TransferHttpCacheModule,
      AppRoutingModule,
      BrowserAnimationsModule,
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

#### server.js
```ts
const express = require('express');
const { Engine } = require('@nguniversal/common/clover/server');
const { join } = require('path');
const { format } = require('url');

const PORT = 8080;
const DIST = join(__dirname, 'dist');

const app = express();
app.set('views', DIST);

app.get('*.*', express.static(DIST, {
  maxAge: '1y',
  fallthrough: false,
}));

// Redirect to default locale
// app.get(/^(\/|\/favicon\.ico)$/, (req, res) => {
//   res.redirect(301, `/en-US${req.originalUrl}`);
// });

const ssr = new Engine();
app.get('*', (req, res, next) => {
  ssr.render({
    publicPath: DIST,
    url: format({
      protocol: req.protocol,
      host: `localhost:${PORT}`,
      pathname: req.path,
      query: req.query,
    }),
    headers: req.headers,
  })
    .then(html => res.send(html))
    .catch(err => next(err));
});

app.listen(PORT, () => {
  console.log(`Node Express server listening on http://localhost:${PORT}`);
});
```

### Running the application
```
ng build
node server.js
```

### CODE REVIEW

File: integration/clover/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# universal-integration
+
+This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 9.0.0-next.4.
```
Comment (kyliau) [inline integration/clover/README.md:3]: Any reason why v9.0.0-next.4 in particular?

File: modules/common/clover/README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
+# Angular SSR project Clover
+
+- What if Angular SSR didn't require the complex `@nguniversal` and `@angular/platform-server` packages on boarding?
```
Comment (kyliau) [inline modules/common/clover/README.md:3]:
```suggestion
- What if Angular SSR didn't require the complex `@nguniversal` and `@angular/platform-server` packages for onboarding?
```

**Review by kyliau — APPROVED** [review]

**Comment by kyliau** [comment]
Target should be `major`?

**Comment by alan-agius4** [comment]
> Target should be `major`?

Unless it's a breaking change, any new feature should be marked with `minor`, `minor` is equivalent to `master-only`.
See: https://github.com/angular/angular/blob/master/docs/TRIAGE_AND_LABELS.md#pr-target

File: modules/common/clover/server/src/server-engine.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        userAgent: options.headers?.['user-agent'] as string | undefined,
+        beforeParse: window => {
+          window.ngDOMRenderMode = true;
+        },
+      });
+
+      const doc = dom.window.document;
+      // tslint:disable-next-line: no-shadowed-variable
+      const ngDOMRenderMode = await new Promise<NGDOMRenderModeAPI>(resolve => {
+        const interval = setInterval(() => {
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply modules/common/clover/server/src/server-engine.ts:76]: Good catch

File: modules/common/clover/README.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  @NgModule({
+    declarations: [
+      AppComponent
+    ],
+    imports: [
+-     BrowserModule,
++     BrowserModule.withServerTransition({
++      appId: 'myapp',
++     }),
++     DomRenderingModule.forRoot(),
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply modules/common/clover/README.md:52]: Yes

**Pull request merged**